doc/builders/builders.chapter.md

```diff
@@ -0,0 +1,56 @@
+# The builders {#chap-builders}
+
+A builder is a function that produces derivations.
+
+:::{.warning}
+This is not to be confused with the [`builder` argument of the Nix `derivation` primitive](https://nixos.org/manual/nix/unstable/language/derivations.html), which refers to the executable that produces the build result, or [remote builder](https://nixos.org/manual/nix/stable/advanced-topics/distributed-builds.html), which refers to the remote build machines.
+:::
+
+Such a function is usually designed to follow a specific workflow, so that one can produce a package by setting a limited set of options relevant to the particular use case instead of using the `derivation` function directly.
+
+[`stdenv.mkDerivation`](#part-stdenv) is the most widely used builder, and serves as a basis for many other builders.
+In addition, it offers various options to customize parts of the builds.
+
+There is no uniform interface for builders.
+[](#chap-trivial-builders) and [fetchers](#chap-pkgs-fetchers) have various input types for convenience, but [language- or framework-specific builders](#chap-language-support) tend to follow the style of `stdenv.mkDerivation`, which accepts an attribute set or a fixed-point function taking an attribute set.
+
+## Function-based attribute recursion support in builders {#sec-build-helpers-finalattrs}
+
+As mentioned above, `stdenv.mkDerivation` could alternatively accept a fixed-point function. The input of such function, typically named `finalAttrs`, is expected to be the final state of the attribute. A builder like this is said to support **function-based attribute recursion**.
+
+As [`stdenv.mkDerivation`'s support of function-based attribute recursion](#mkderivation-recursive-attributes) was first included in the Nixpkgs release 22.05, `mkDerivation`-style builders don't necessarily come with such support.
+
+The Nixpkgs library function `lib.customisation.extendMkDerivation` can be used to define a builders with recursive attributes support from an existing one.
+
+### Example {#ssec-build-helpers-finalattrs-example}
```

Comment:
Thank you for telling me that!

BTW, this part of documentation seems to be a duplication of the function reference of `lib.extendMkDerivation`. I originally write it here just because `lib.customization` was not included in the function reference then.

I think we should either:
1.  Link from here to the function reference, or
2.  Link form the function reference to here.

The first one sounds more sensible to me. It's a pity that we don't have a way to link against the documentation of each function.

Reply:
> This should use an `:::{example}` block, but I wouldn't... block on that.

Addressed.

The example block seems to contain rendering bug that (randomly) affects the style of the next few paragraphs (e.g. showing underline when moving the point er across the paragraphs).

Reply:
> It's a pity that we don't have a way to link against the documentation of each function.

Update:

We do have a way to link against it. The anchor would be `finction-library-lib.customisation.extendMkDerivation`. I really hope these automatic generation of anchors could be mentioned somewhere in the documentation.